node_modules/oae-authentication/lib/api.js

```diff
@@ -677,6 +679,113 @@ var getUserIdFromLoginId = module.exports.getUserIdFromLoginId = function(tenant
+
+            // The secret is stored for 24 hours
+            Cassandra.runQuery('UPDATE "AuthenticationLoginId" USING TTL 86400 SET "secret" = ? WHERE "loginId" = ?', [secret, _flattenLoginId(loginId)], function(err) {
+                if (err) {
+                    log().error({'err': err}, 'Error in creating a secret token');
+                    return callback(err);
+                }
+
+                // Generate the email content with secret token
+                var emailData = {
+                    'tenant': ctx.tenant(),
+                    'user': user,
+                    'username': username,
+                    'baseUrl': TenantsUtil.getBaseUrl(ctx.tenant()),
+                    'skinVariables': require('oae-ui').getTenantSkinVariables(ctx.tenant().alias),
+                    'secret': secret,
+                };
+
+                // Send the email to the user
+                EmailAPI.sendEmail('oae-authentication', 'reset', user, emailData, null, callback);
+            });
+        });
+    });
+    
+};
+
+/**
+ * Reset the password for an user with an existing username and a valid token
+ *
+ * @param  {Context}    ctx            Standard context object containing the current user and the current tenant
+ * @param  {String}     tenantAlias    The tenant of the login id
+ * @param  {String}     userName       The user's own unique username 
+ * @param  {String}     token          The token that user generated within 24 hours
+ * @param  {String}     newPassword    The new password for the user
+ * @param  {Function}   callback       Standard callback function
+ * @param  {Object}     callback.err   An error that occurred, if any
+ */
+var resetPassword = module.exports.resetPassword = function(ctx, tenantAlias, username, secret, newPassword, callback){
+	
+	  // Associate the userID with loginID for later operations
+    tenantAlias = tenantAlias || ctx.tenant().alias;
+
+    // Create the loginid object first, so it can be passed into the validation
+    var loginId = new LoginId(tenantAlias, AuthenticationConstants.providers.LOCAL, username);
+	
+	  // Lookup the database to check whether a token is associated with the current
+	  Cassandra.runQuery('SELECT "userId", "secret" FROM "AuthenticationLoginId" WHERE "loginId" = ?', [_flattenLoginId(loginId)], function(err, rows) {
+        
+        if (err) {
+            return callback(err);
+        } else if (_.isEmpty(rows)) {
+            // No user found with that login id
+            return callback({'code': 401, 'msg': 'No password found for this principal'});
+        }
+
+        // Get the secret column out of the row
+        var dbSecret = rows[0].get('secret');
+
+        // If the secret column was not found or its value didn't match, something is wrong.
+        if (!dbSecret || dbSecret.value !== secret) {
+            log().warn({'loginId': loginId}, 'Potential hacking attempt detected');
```

Comment:
This logging message would be more useful if it had more information about what actually happened - there are many reasons why a secret could not be found that are not related to hacking (for example a user clicked on an old link).